### ISSUE TITLE
DisplayObjectContainer getBounds should ignore the invisible children

### ISSUE THREAD

**Issue by leminhduc** [Gamua/Starling-Framework#765]
Hi Daniel,

I think there might be a bug in DisplayObjectContainer getBounds function. It loops through all the children to calculate the bound. However, if a child is invisible, the child's bound should be ignore. getBounds should check for child's visibility.

For example: I have a Sprite A with 2 children  { child1: { x: 0, y: 0, width:10, height:10}, child2: { x: 10, y: 0, width:10, height:10} }

A.child1.visible = A.child2.visible = true;
A.width == 20  // => correct 

A.child1.visible = A.child2.visible = false;
A.width == 20  // => wrong , it should be equal to 0.

Please explain if I am wrong. Thank you very much
David

**Comment by gsynuh** [comment]
Hi @leminhduc,

Just a quick reaction before you get a better answer from anyone else. 

Your comment is interesting, but to me getBounds is not about visibilty. if it was, how about objects of alpha = 0, or objects with transparent textures or with transparent areas? 

No matter what Starling is actually doing to the geometry behind the scenes (not rendering it at all I guess) to me as long as that object is a child of another, no matter what its other properties are, all of its transformation and 'spatial properties' should exist in all other calculations from parents in the hierarchy as the geometry should be here regardless of the rendered result.

Then the fact that touch events go through invisible object is more about interaction, one could argue that it shouldn't perhaps, using the same argument I made ... But that's how it is as in interaction I expect my user not to know there's something there anyway.

- the short version is that's how things are conventionally, it could've been different but I would've called the function getVisualBounds then and have alpha be considered and return 0 if the object was out of screen even...

**Comment by leminhduc** [comment]
Hi @gsynuh, 

Thank you very much for your fast response. I agree with your point below:

"all of its transformation and 'spatial properties' should exist in all other calculations from parents in the hierarchy as the geometry should be here regardless of the rendered result."

However, this is about object's width and height. I believe that width and height are tight strongly with the rendered result. That's why these function in DisplayObject:

public function get width():Number { return getBounds(mParent, sHelperRect).width; }
public function get height():Number { return getBounds(mParent, sHelperRect).height; }

always recalculate the width and height based on its current rendered state (i.e properties, children...). If it is independent from rendered result, should it be stored in a private variable similar to mX, mY ?

Perhaps I do not want to argue about Starling design, I just want to solve my current problem with width and height: 

Every frame I want to align Object A to screen center. And its children's visibility is changing sometimes. I expect Object A is visually at the center all the time. How should I do it properly? In native Flash, it is simple as this : A.x = (screenWidth - A.width) * 0.5.

**Comment by gsynuh** [comment]
sure width and height and rendered results are linked in a way, but what affects width and height is the base width and height of the geometry, and the transformation matrix - that transformation matrix can change all the time so that's why parents have to recalculate bounds (and so width/height) all the time to be sure.

Please take your questions to the Starling forums (and post the link to the topic back here)

**Comment by leminhduc** [comment]
Hi @gsynuh,

Wow you are so fast!  Yes i was wrong about native flash, Starling and native flash behave similarly. So it is not a bug, let me close this thread. 

I post question here, hopefully someone can share knowledge about it.
http://forum.starling-framework.org/topic/should-displayobjectcontainer-getbounds-ignore-the-invisible-children

Thank you very much

**Comment by PrimaryFeather** [comment]
Yes, it's possible to argue either way — sometimes, it would be helpful to have the bounds just report the visible elements; and sometimes, it's the other way round. When I wrote the code, I can remember I asked myself the same question, and then settled it by having a look at how native Flash does it. 

When you look at Starling's "hitTest" method, you'll see that I ran into the exact same issue: should it be a hit if the object is invisible? That's why I added a "forTouch" parameter to that method: if that's false, it's supposed to ignore object visibility. However, it turned out that this was rather useless: a hitTest is almost exclusively done for touch / mouse interaction. So I'll probably remove that parameter in the future.

In any case, thanks to both of you for the discussion! It's always important to question such things. After working on and with Starling for such a long time, I tend to develop blind spots in such areas. :wink:
